### PR TITLE
minimal SMIE setup for indentation

### PR DESCRIPTION
**Pull request by tko**
This is intentionally discarding the grammar definition in smie branch as it seems to interfere with indentation making it require quite a few adjustments that aren't really making much sense. I think you'll need the grammar for sexp navigation and manipulation, but for indentation this almost works.

I haven't figured out how to get the colon show up in puppet-smie-rules which should be enough to fix the problem with resource indentation but I suspect it needs some grammar/tokenizer changes.

### CODE REVIEW

**Comment by lunaryorn** [comment]
@tko Thanks, that looks like a great piece of work.  Will take a look and hopefully merge soon.

File: puppet-mode.el [hunk 1]
```diff
@@ -632,6 +633,40 @@ of the initial include plus puppet-include-indent."
         (indent-line-to 0)))))
 
 
+;;;; SMIE
+
+(defvar puppet-use-smie nil
+  "Whether to use SMIE or not.")
```
Comment (lunaryorn) [inline puppet-mode.el]: I don't think that we need this.  If SMIE works for us, we'll just require an Emacs version that's recent enough.  But it's too much effort to maintain to different indentation/navigation engines in this mode, imho.

Comment (tko) [inline puppet-mode.el]: The problem is that it doesn't really work so it should be opt-in and flipping a variable is easier than switching branches. Eventually, if it ever works, sure you don't need a variable, but until then...

File: puppet-mode.el [hunk 2]
```diff
@@ -1093,7 +1128,11 @@ for each entry."
   (setq-local beginning-of-defun-function #'puppet-beginning-of-defun-function)
   ;; Indentation
   (setq-local indent-line-function #'puppet-indent-line)
-  (setq indent-tabs-mode puppet-indent-tabs-mode)
+  (if puppet-use-smie
+      (smie-setup puppet-smie-grammar #'puppet-smie-rules
+                  :forward-token #'puppet--smie-forward-token
+                  :backward-token #'puppet--smie-backward-token)
+      (setq indent-tabs-mode puppet-indent-tabs-mode))
```
Comment (lunaryorn) [inline puppet-mode.el]: Why did you move this into the `if`?  `indent-tabs-mode` should be independent of SMIE shouldn't it?

File: puppet-mode.el (hunk 1 above)
Comment (lunaryorn) [inline puppet-mode.el]: I see.  Let's keep it then.

File: puppet-mode.el (hunk 2 above)
Comment (tko) [inline puppet-mode.el]: I have no idea. That's what 5efe0e8787fdd9afacf2afd3e9b5aefd6cc1e08f did.

Comment (lunaryorn) [inline puppet-mode.el]: I see, but that was likely a mistake I made back then.  I think you can move it out of `if`, and replace the `if` with a `when`.

File: puppet-mode.el (hunk 1 above)
Comment (bbatsov) [inline puppet-mode.el]: The problem is that the current indentation doesn't work very well either. It's better to focus just on one. Perhaps we shouldn't merge this until you get it into some "good enough" state.

File: puppet-mode.el [hunk 3]
```diff
@@ -632,6 +633,40 @@ of the initial include plus puppet-include-indent."
...
+   (smie-bnf->prec2
+    '((id)))))
+
+(defun puppet-smie-rules (kind token)
+  "Indentation rules for Puppet."
+  (when puppet-smie-verbose
+    (save-excursion
+      (let ((prev (smie-indent-backward-token))
+            (next (smie-indent-forward-token)))
+        (message "kind=%S token=%S prev=%S next=%S parent=%S"
```
Comment (bbatsov) [inline puppet-mode.el]: message?

**Comment by bbatsov** [comment]
Not sure which mode you used for inspiration. I think these days `elixir-mode` and `swift-mode` are some of the best examples of smie usage.

File: puppet-mode.el (hunk 3 above)
Comment (tko) [inline puppet-mode.el]: It prints things. I don't really do elisp so I'm not familiar with how you normally trace things. *shrug*

**Comment by tko** [comment]
I'm trying to do as little as possible to convince SMIE to indent the code properly and if it weren't for the resources titles I think the do-nothing implementation would work nicely but alas. I've tried looking at ruby and other modes but they seem rather complex doing things I don't really understand or much care about :-/

**Comment by tko** [comment]
Oh, when trying to fix indentation problems it would be helpful to have a range of tests to see you're not accidentally breaking everything else in the process. SMIE seems to be fairly sensitive about changes. Having a bunch of sample files and checking reindenting doesn't change anything would be nice.

**Comment by tko** [comment]
FWIW I tried adding some half baked test setup for indentation in https://github.com/tko/puppet-mode/commit/90651768d85f7ff4c657b8401df7a321eafaaa90

**Comment by apchamberlain** [comment]
I'm willing to help with the test suite, although I have limited free time.  I used to work at Puppet Labs in QA and have a good variety of sample code available from which to glean representative cases.

**Comment by tko** [comment]
@apchamberlain I've some tests in https://github.com/tko/puppet-mode/commit/90651768d85f7ff4c657b8401df7a321eafaaa90 where you can more or less just drop in new properly indented files and the tests will just check nothing changes. It needs some finishing and decide how to merge it - the number of passing tests is limited with the code in master.

As of now this PR passes almost all the tests, but my confidence on it is only as good as the test coverage.

**Comment by tko** [comment]
Anything?

**Comment by lunaryorn** [comment]
@tko Sorry, forgot about this.

File: puppet-mode.el (hunk 3 above)
Comment (lunaryorn) [inline puppet-mode.el]: Please remove it for merging.

File: test/puppet-mode-test.el [hunk 4]
```diff
@@ -601,6 +601,82 @@ package { 'bar':
     (should (local-variable-p 'imenu-create-index-function))
     (should (equal imenu-create-index-function #'puppet-imenu-create-index))))
 
+
+;;;; Indentation
+
+(defmacro flycheck-ert-with-temp-buffer (&rest body)
```
Comment (lunaryorn) [inline test/puppet-mode-test.el]: flycheck?

File: puppet-mode.el (hunk 3 above)
Comment (tko) [inline puppet-mode.el]: Just to be clear, you did notice it's wrapped in a conditional one needs to manually enable for anything to be printed, right? I'm beginner in elisp and not so sure how to assist with debugging properly, thought it would be helpful for the next person.

File: test/puppet-mode-test.el (hunk 4 above)
Comment (tko) [inline test/puppet-mode-test.el]: Yes, copypasta.

I've never user ert (or much of elisp) before and with little educated googling those seemed useful for the kinds of testing setup I thought was appropriate here.

Comment (lunaryorn) [inline test/puppet-mode-test.el]: @tko It's appropriate, but you should rename the functions, from `flycheck-ert-*` to `puppet-test-*` or something similar.  This just isn't Flycheck, after all :smiley:

Comment (tko) [inline test/puppet-mode-test.el]: If you say so, I just thought to give credit where it's due (or rather since I'm lazy I just didn't want to touch a thing.) Guess a comment will suffice here.

File: puppet-mode.el (hunk 3 above)
Comment (lunaryorn) [inline puppet-mode.el]: Yes, but it's still just redundant debugging code which serves no specific purpose by itself, and also only covers the particular instance of debugging that you needed in the past.  But it's of no use generally, so please remove it and the variable as well.  

If anyone needs to debug this particular place again, they can easily add in their own debugging code again and press `C-M-x`.

**Comment by rski** [comment]
@tko You have done some good work here, would you be interested in picking it up again?

**Comment by tko** [comment]
@rski unfortunately not really. We got other people taking care of puppet so my incentives rather disappeared. And I've about already forgotten everything about elisp and especially smie by now.

**Comment by rski** [comment]
ok fair enough, it has been quite some time. Thanks for all your work, I will pick it up then from here (and ofc you'll be properly attributed)

**Comment by bbatsov** [comment]
I think it's time to close this.

**Comment by rski** [comment]
one day SMIE's day will come :< (although i have been toying with the idea of a full parser a la js2-mode)
